cli/azd/pkg/async/task.go

```diff
@@ -0,0 +1,149 @@
+package async
+
+// Task represents a long running async operation
+type Task[R comparable] struct {
+	isComplete    bool
+	hasResult     bool
+	result        R
+	resultChannel chan R
+	error         error
+}
+
+// Runs the specified taskFn as a go routine
+func (t *Task[R]) Run(taskFn TaskRunFunc[R]) {
+	go func() {
+		context := NewTaskContext(t)
+
+		taskFn(context)
+		t.complete(context.result, context.error)
+	}()
+}
+
+// Checks whether or not the Task has completed
+func (t *Task[R]) IsCompleted() bool {
+	return t.isComplete
+}
+
+// Waits for a result to become available and returns the task value
+func (t *Task[R]) Result() R {
+	if t.hasResult {
+		return t.result
+	}
+
+	t.result = <-t.resultChannel
+	t.hasResult = true
+
+	return t.result
+}
+
+// Awaits the async execution and returns the result and error status
+func (t *Task[R]) Await() (R, error) {
+	result := t.Result()
+	return result, t.error
+}
+
+// Marks the current task as complete and sets internal error/result state and cleans up channels
+func (t *Task[R]) complete(result R, err error) {
+	t.error = err
+	if t.error == nil {
+		t.resultChannel <- result
+	}
+
+	t.isComplete = true
+	close(t.resultChannel)
+}
+
+// Creates a new instance of a Task
+func NewTask[R comparable]() *Task[R] {
+	return &Task[R]{
+		resultChannel: make(chan R, 1),
+	}
+}
+
+// Creates and schedules the task function and returns a task instance that holds the future result
+func RunTask[R comparable](taskFn TaskRunFunc[R]) *Task[R] {
+	task := NewTask[R]()
+	task.Run(taskFn)
+
+	return task
+}
+
+// Represents an async Task operation that support progress reporting
+type TaskWithProgress[R comparable, P comparable] struct {
+	Task[R]
+	progressChannel chan P
+}
+
+// Gets the go channel that represents the task progress
+func (t *TaskWithProgress[R, P]) Progress() <-chan P {
+	return t.progressChannel
+}
+
+// Runs the specified taskFn as a go routine
+func (t *TaskWithProgress[R, P]) Run(taskFn TaskWithProgressRunFunc[R, P]) {
+	go func() {
+		context := NewTaskContextWithProgress(t)
+
+		taskFn(context)
```

Comment:
Makes me a little worried that we are creating a circular reference here with the `context` as it contains a pointer to the task that is running it, so within `taskFn` I can call `context.task.Run()` ??  
Can you help me understand why do we need to pass the reference to the task to the function running in another routine?

Reply:
`context.task` wouldn't be available since it is a private member not available to the other packages.  The `TaskContext` only has functions to set results, errors & progress.